### ISSUE TITLE
Where to do complex aggregation and data processing?

### ISSUE THREAD

**Issue by vctls** [stwe/DatatablesBundle#653]
Excuse me if this is not the right place to ask, I didn't know where else to ask, as this is kind of specific.

tl;dr Is this bundle actually fit for heavily processed data and complex requests?

I sometimes need to do some complex data processing before displaying it, and I am not sure how I can do it while using this bundle.
One simple example:
I have a Person entity, each person has many Files, and each File can have a different State.
In the Person datatable, I need to add a "File states" column, which displays a string aggregation of the distinct States of the Files of each person.
Like this:

| Name | Files | File states                     |
| ------ | ---- | ------------------------ |
| Ann    | 3     | pending, accepted         |
| Paul    | 1     | accepted                       |
| John   | 5     | new, pending, accepted |

In can manage to get a list of States using the `data` option.
In version 0.13, I used the (now gone?) `render` option to reduce the list client-side, so each State appears only once.
But I don't see how I can do that using a DQL expression.

Anyway, all of these methods feel really "hacky" to me. If I got the Symfony best practices right, I should be doing any complex queries in the entities' EntityRepository.
With this bundle, we're skipping the EntityRepository alltogether.

So, Is there a recommended way of doing this kind of things with the tools provided in the bundle?
Or should I use something else?

**Comment by stwe** [comment]
The query is created automatically. But you can change it

https://github.com/stwe/DatatablesBundle/blob/master/Resources/doc/query.md

```
    if ($isAjax) {
        $responseService = $this->get('sg_datatables.response');
        $responseService->setDatatable($datatable);

        $datatableQueryBuilder = $responseService->getDatatableQueryBuilder();
        $datatableQueryBuilder->buildQuery();

        // Do something with the automatically created query:
        // dump($datatableQueryBuilder->getQb()->getDQL()); die();

        return $responseService->getResponse();
    }
```

**Comment by vctls** [comment]
Ok, so the idea is to get the QueryBuilder and change whatever I want.
I suppose I will need to carry it around if I want to put my logic somewhere else.
I'll give it a try.
Thank you!
